ext/liquid_c/variable_lookup.c

```diff
@@ -0,0 +1,64 @@
+#include "liquid.h"
+#include "context.h"
+
+static ID id_has_key, id_aref, id_fetch;
+static ID id_ivar_lookups, id_ivar_name, id_ivar_command_flags;
+
+VALUE variable_lookup_evaluate(VALUE self, VALUE context)
+{
+    VALUE command_flags = 0;
+    VALUE name = rb_ivar_get(self, id_ivar_name);
+    name = context_evaluate(context, name);
+
+    VALUE object = context_find_variable(context, name, Qtrue);
+
+    VALUE lookups = rb_ivar_get(self, id_ivar_lookups);
+    Check_Type(lookups, T_ARRAY);
+
+    for (long i = 0; i < RARRAY_LEN(lookups); i++) {
+        VALUE key = context_evaluate(context, RARRAY_AREF(lookups, i));
+        VALUE next_object;
+
+        if (rb_respond_to(object, id_aref) && (
+            (rb_respond_to(object, id_has_key) && rb_funcall(object, id_has_key, 1, key)) ||
+            (rb_obj_is_kind_of(key, rb_cInteger) && rb_respond_to(object, id_fetch))
+        )) {
+            next_object = rb_funcall(object, id_aref, 1, key);
+            next_object = materialize_proc(context, object, key, next_object);
```

Comment:
should we have a `lookup_and_evaluate` function instead of having to repeat this `[]`+`materialize_proc` pattern in multiple places? dunno.

Reply:
I actually find this clearer, the code at this point in Ruby ends up having the comment `# if its a proc we will replace the entry with the proc` which is now simply written here. It also gets a bit messy to combine the logic with find_variable. Though, it might be a good idea so we can get hash-specific optimizations into this path too. I didn't bother originally since at this point you're mostly dealing with a drop instead.